### PR TITLE
Minor external array cleanups, suggested by Michael

### PR DESCRIPTION
**Pull request by lydia-duncan**
Add explicit return type to convertToExternalArray, to avoid confusion.

Also permit ExternArray to store empty arrays.  Empty arrays are
represented by the domain {1..0} in Chapel, but 1 is not normally permitted
as a lower bound for an external array.  Since we can't modify the array
bounds after it has been created, it is okay to allow the empty array to be
representable with an ExternArray.

Add a test to lock in the support for empty arrays 

Passed a full paratest with futures

### CODE REVIEW

File: modules/internal/ExternalArray.chpl [hunk 1]
```diff
@@ -55,7 +55,8 @@ module ExternalArray {
         halt("there should only be one set of indices, not multiple dimensions");
       }
       var r = inds(1);
-      if (r.low != 0) {
+      // Allow empty arrays (which have the domain 1..0)
```
Comment (mppf) [inline modules/internal/ExternalArray.chpl]: 0..-1 is just as reasonable for an empty array.
It seems to me that if you want to check this, you'd check for r.low < r.high.

Comment (mppf) [reply modules/internal/ExternalArray.chpl]: I'm not personally sold that these need to support 1..0 bounds, but I'm also OK with it.

File: modules/internal/ExternalArray.chpl [hunk 2]
```diff
@@ -55,7 +55,8 @@ module ExternalArray {
...
+      if (r.low != 0 && !(r.low == 1 && r.high == 0)) {
         halt("external arrays always have a lower bound of 0");
```
Comment (mppf) [inline modules/internal/ExternalArray.chpl]: Should it say non-empty external arrays have a lower bound of 0?
Maybe do one more check that we're not relying on the lower bound.

File: modules/internal/ExternalArray.chpl (hunk 1 above)
Comment (lydia-duncan) [reply modules/internal/ExternalArray.chpl]: We were running into this problem because `0..#0` was getting transformed into `1..0` (https://github.com/chapel-lang/chapel/issues/5165 tracks the desire to fix that issue).  I was using `0..#0` to create the array when we know the size, and it is a lot more work for a user to create an external array with their own domain (though if they do, {0..-1} should be accepted without additional work right now).

I don't feel strongly about whether `2..1` or any other empty domain gets supported, but I think I'd rather wait until there was a strong desire to do so.

File: modules/internal/ExternalArray.chpl (hunk 2 above)
Comment (lydia-duncan) [reply modules/internal/ExternalArray.chpl]: I'm a little confused, can you clarify what you're thinking here?

Comment (mppf) [reply modules/internal/ExternalArray.chpl]: 1..0 bounds would mean the lower bound could be 0, couldn't it?

Comment (lydia-duncan) [reply modules/internal/ExternalArray.chpl]: Yes.  I think the error message change makes sense.  Should have clarified, I'm mostly confused about the "extra check" portion of that comment

Comment (mppf) [reply modules/internal/ExternalArray.chpl]: > Maybe do one more check that we're not relying on the lower bound.

Is there code in this module that won't work if r.low == 1 and the array is actually nil?

Comment (lydia-duncan) [reply modules/internal/ExternalArray.chpl]: I don't think so.  The rest of the implementation seems to only check on the size of the domain, which is stored separately instead of as a range.  This code doesn't pass the range on, just its size.

It looks like ExternDom's domRange method returns 0..#0/1..0 (which dsiGetIndices and dsiDim(s) relies on) but seems to be used in loops where an empty range would just mean the loop got skipped anyways

Comment (lydia-duncan) [reply modules/internal/ExternalArray.chpl]: That said, I don't fully understand the purpose of domRange/dsiGetIndices/dsiDim/dsiDims, so it's possible there is something I'm missing

Comment (lydia-duncan) [reply modules/internal/ExternalArray.chpl]: Circling back now that I am on top of things again, do you think there is an additional action item for this PR or is it good to go?  (I know there's a green approval check mark but this discussion made me uncertain)

Comment (mppf) [reply modules/internal/ExternalArray.chpl]: I'm OK with it, just wanted you to think about it, which you did

**Comment by lydia-duncan** [comment]
(passed paratest after master merge, merging now)
